central/graphql/resolvers/node_scan.go

```diff
@@ -29,5 +37,55 @@ func (resolver *nodeScanResolver) NodeComponentCount(_ context.Context, args Raw
 }
 
 func (resolver *nodeScanResolver) NodeComponents(_ context.Context, args PaginatedQuery) ([]NodeComponentResolver, error) {
-	return resolver.root.NodeComponents(resolver.ctx, args)
+	query, err := args.AsV1QueryOrEmpty()
+	if err != nil {
+		return nil, err
+	}
+	return getNodeComponentResolvers(resolver.ctx, resolver.root, resolver.data, query)
+}
+
+func getNodeComponentResolvers(ctx context.Context, root *Resolver, nodeScan *storage.NodeScan, query *v1.Query) ([]NodeComponentResolver, error) {
+	query, _ = search.FilterQueryWithMap(query, mappings.NodeComponentOptionsMap)
+	predicate, err := nodeComponentPredicateFactory.GeneratePredicate(query)
+	if err != nil {
+		return nil, err
+	}
+
+	idToComponent := make(map[string]*nodeComponentResolver)
+	for _, embeddedComponent := range nodeScan.GetComponents() {
+		if !predicate.Matches(embeddedComponent) {
+			continue
+		}
+
+		os := nodeScan.GetOperatingSystem()
+		id := scancomponent.ComponentID(embeddedComponent.GetName(), embeddedComponent.GetVersion(), os)
+		if _, exists := idToComponent[id]; !exists {
+			component := common.GenerateNodeComponent(os, embeddedComponent)
+			resolver, err := root.wrapNodeComponent(component, true, nil)
+			if err != nil {
+				return nil, err
+			}
+			resolver.ctx = embeddedobjs.NodeComponentContext(ctx, os, nodeScan.GetScanTime(), embeddedComponent)
+			idToComponent[id] = resolver
+		}
+	}
+
+	// For now, sort by IDs.
+	resolvers := make([]*nodeComponentResolver, 0, len(idToComponent))
```

Comment:
nit: I think naming these so there is more distinction between resolvers and resolverI would be useful. Also resolverI should probably be plural.